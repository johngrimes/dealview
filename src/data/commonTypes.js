export const AddressDefaults = {
  line1: '',
  line2: '',
  line3: '',
  locality: '',
  state: '',
  postcode: '',
}

export const DateDisplayFormat = 'D MMM YYYY'
export const DateStorageFormat = 'YYYY-MM-DD'

export const parseNumber = string =>
  typeof string === 'string' ? parseInt(string) : undefined

export const formatNumber = number => {
  if (typeof number !== 'number') return undefined
  const numString = number.toString()
  return [...numString].reduce((prev, curr, i, arr) => {
    return arr.length - i - 1 !== 0 && (arr.length - i - 1) % 3 === 0
      ? prev + curr + ','
      : prev + curr
  }, '')
}

export const formatDollars = number =>
  typeof number === 'number' ? '$' + formatNumber(number) : undefined

export const unformatNumber = string => {
  return typeof string === 'string'
    ? parseInt(string.replace(/,/g, ''), 10)
    : undefined
}

export const unformatDollars = string => {
  return typeof string === 'string'
    ? unformatNumber(string.replace(/\$/g, ''))
    : undefined
}
