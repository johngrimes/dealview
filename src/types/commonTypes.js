// @flow

import type { Dispatch } from 'redux'

export type Thunk = (dispatch: Dispatch) => any

export type ObjectMap = { [id: string]: Object }
export type ObjectWithId = Object & { +id: string }

export type Address = {
  +line1: string,
  +line2: string,
  +line3: string,
  +locality: string,
  +state: string,
  +postcode: string,
}

export const AddressDefaults = {
  line1: '',
  line2: '',
  line3: '',
  locality: '',
  state: '',
  postcode: '',
}

export const DateDisplayFormat = 'D MMM YYYY'
export const DateStorageFormat = 'YYYY-MM-DD'

export const formatNumber = (number: number): string => {
  const numString = number.toString()
  // The Array.from is a workaround for https://github.com/facebook/flow/issues/1059, and can be removed once this is resolved.
  return [...Array.from(numString)].reduce((prev, curr, i, arr) => {
    return (arr.length - i - 1) !== 0 && (arr.length - i - 1) % 3 === 0
      ? prev + curr + ','
      : prev + curr
  }, '')
}

export const formatDollars = (number: number): string => '$' + formatNumber(number)
