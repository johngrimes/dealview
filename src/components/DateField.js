import React from 'react'
import DatePicker from 'react-datepicker'
import moment from 'moment'

import { DateDisplayFormat, DateStorageFormat } from '../data/commonTypes.js'

import 'react-datepicker/dist/react-datepicker-cssmodules.css'
import './styles/DateField.css'

class DateField extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      value: props.value,
      touched: false,
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
  }

  handleChange(value) {
    this.setState(
      () => ({ value: value, touched: true }),
      () => {
        if (this.props.onChange) this.props.onChange(value)
      }
    )
  }

  handleFocus(event) {
    const target = event.target
    if (this.props.onFocus) {
      this.props.onFocus(target.name)
    }
  }

  componentWillReceiveProps(props) {
    this.setState(() => ({ value: props.value }))
  }

  render() {
    const {
      name,
      label,
      errors,
      forceErrorDisplay,
      minDate,
      maxDate,
    } = this.props
    const { value, touched } = this.state

    const labelTag = label
      ? <label htmlFor={name}>
          {label}
        </label>
      : null
    const errorTags = []
    if (errors) {
      errors.forEach((msg, i) =>
        errorTags.push(
          <div key={i} className='error'>
            {msg}
          </div>
        )
      )
    }
    const inputClass = touched && errorTags.length > 0 ? 'with-errors' : ''
    let selected
    if (typeof value === 'string' && value.length > 0) {
      selected = moment(value, DateStorageFormat)
    } else if (value instanceof moment) {
      selected = value
    }

    return (
      <div className='control-group'>
        {labelTag}
        <div className='date-picker'>
          <DatePicker
            name={name}
            dateFormat={DateDisplayFormat}
            className={inputClass}
            selected={selected}
            showYearDropdown
            autoFocus={this.props.focus === name}
            onFocus={this.handleFocus}
            minDate={minDate}
            maxDate={maxDate}
            onChange={moment =>
              this.handleChange(moment.format(DateStorageFormat))}
          />
        </div>
        {(forceErrorDisplay || touched) &&
          errorTags.length > 0 &&
          <div className='errors'>
            {errorTags}
          </div>}
      </div>
    )
  }
}

export default DateField
