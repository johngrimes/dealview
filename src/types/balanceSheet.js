// @flow

import _ from 'lodash'
import moment from 'moment'
import type { Moment } from 'moment'

import { DateStorageFormat } from 'types/commonTypes'
import { getValuationAtDate } from 'types/assets/asset'
import type { Asset, AssetMap } from 'types/assets/asset'
import type { LiabilityMap } from 'types/liabilities/liability'

export type BalanceSheet = {
  +totalAssets: number,
  +totalLiabilities: number,
  +equity: number
}
export type BalanceSheetOverTime = { +[date: string]: BalanceSheet }

export const balanceSheetOverTime = (assets: AssetMap, liabilities: LiabilityMap,
                              startDate: Moment, endDate: Moment): BalanceSheetOverTime => {
  // The Array.from is a workaround for https://github.com/facebook/flow/issues/1059, and can be removed once this is resolved.
  return Object.assign({}, ...Array.from(calcBalanceSheet(assets, liabilities, startDate, endDate)))
}

function* calcBalanceSheet(assets: AssetMap, liabilities: LiabilityMap,
                           date: Moment, endDate: Moment): Generator<BalanceSheetOverTime, void, void> {
  let nextDate = date
  while (!nextDate.isAfter(endDate)) {
    const totalAssets = sumAssetValueAtDate(_.values(assets), nextDate)
    const totalLiabilities = 0  // sumLiabilityValueAtDate(liabilities, date)
    yield {
      [nextDate.format(DateStorageFormat)]: {
        totalAssets,
        totalLiabilities,
        equity: (totalAssets - totalLiabilities),
      },
    }
    nextDate = moment(nextDate).add(1, 'days')
  }
}

const sumAssetValueAtDate = (assets: Asset[], date: Moment): number => {
  if (assets.length === 0) return 0
  const { head, tail } = { head: _.head(assets), tail: _.tail(assets) }
  return getValuationAtDate(head, date) + sumAssetValueAtDate(tail, date)
}
