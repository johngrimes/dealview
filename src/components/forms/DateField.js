// @flow

import React from 'react'
import DatePicker from 'react-datepicker'
import moment from 'moment'
import type { Moment } from 'moment'

import { DateDisplayFormat, DateStorageFormat } from 'types/commonTypes'
import type { FieldErrors } from 'utils/FormValidation'

import 'react-datepicker/dist/react-datepicker-cssmodules.css'

type Props = {
  +name: string,
  +value?: string | Moment,
  +label?: string,
  +errors?: FieldErrors,
  +forceErrorDisplay?: boolean,
  +focus?: string,
  +onChange?: (value: string) => void,
  +onFocus?: (fieldName: string) => void
}

type State = {
  +value?: string,
  +touched: boolean
}

class DateField extends React.Component {
  props: Props
  state: State
  handleChange: (value: string) => void
  handleFocus: (event: { target: { name: string } }) => void

  constructor(props: Props) {
    super(props)
    this.state = {
      value: props.value,
      touched: false,
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
  }

  handleChange(value: string): void {
    this.setState(
      () => ({ value: value, touched: true }),
      () => { if (this.props.onChange) this.props.onChange(value) }
    )
  }

  handleFocus(event: { target: { name: string } }): void {
    const target = event.target
    if (this.props.onFocus) {
      this.props.onFocus(target.name)
    }
  }

  componentWillReceiveProps(props: Props) {
    this.setState(() => ({ value: props.value }))
  }

  render() {
    const { name, label, errors, forceErrorDisplay } = this.props
    const { value, touched } = this.state
    const labelTag = label ? <label htmlFor={name}>{label}</label> : null
    const errorTags = []
    if (errors) {
      errors.forEach((msg, i) =>
        errorTags.push(<div key={i} className='error'>{msg}</div>)
      )
    }
    const inputClass = touched && errorTags.length > 0
      ? 'with-errors'
      : ''
    let selected
    if (typeof value === 'string' && value.length > 0) {
      selected = moment(value, DateStorageFormat)
    } else if (value instanceof moment) {
      selected = value
    }

    return (
      <div className='control-group'>
        {labelTag}
        <DatePicker name={name} dateFormat={DateDisplayFormat}
          className={inputClass}
          selected={selected} showYearDropdown
          autoFocus={this.props.focus === name}
          onFocus={this.handleFocus}
          onChange={(moment) => this.handleChange(moment.format(DateStorageFormat))} />
        {(forceErrorDisplay || touched) && errorTags.length > 0 && <div className='errors'>{errorTags}</div>}
      </div>
    )
  }
}

export default DateField
