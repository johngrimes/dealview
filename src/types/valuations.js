// @flow

import moment from 'moment'

import { DateStorageFormat } from 'types/commonTypes'

export type Valuation = {
  +date?: string,
  +amount?: number,
  +note?: string
}
export const ValuationDefault = {
  date: moment().format(DateStorageFormat),
  note: '',
}
export type Valuations = Valuation[]
