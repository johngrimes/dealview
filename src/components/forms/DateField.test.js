/* global expect */

import DateField from 'components/forms/DateField'

import React from 'react'
import { shallow } from 'enzyme'
import DatePicker from 'react-datepicker'
import moment from 'moment'

import { DateStorageFormat } from 'types/commonTypes'

describe('DateField', () => {
  it('should render', () => {
    const props = {
      name: 'someDate',
      value: '1981-07-26',
      label: 'Some Date',
    }
    const wrapper = shallow(<DateField {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should be tolerant of an undefined value prop', () => {
    const props = {
      name: 'someDate',
      label: 'Some Date',
    }
    const wrapper = shallow(<DateField {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should be tolerant of an empty value prop', () => {
    const props = {
      name: 'someDate',
      label: 'Some Date',
      value: '',
    }
    const wrapper = shallow(<DateField {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should display errors if touched', () => {
    const props = {
      name: 'someDate',
      value: '1981-07-26',
      label: 'Some Date',
      errors: [ 'does not look right', 'smells funny' ],
    }
    const wrapper = shallow(<DateField {...props} />)
    wrapper.setState({
      value: '2013-06-10',
      touched: true,
    }, () => {
      expect(wrapper.find('.error')).toHaveLength(2)
      expect(wrapper).toMatchSnapshot()
    })
  })

  it('should not display errors if not touched', () => {
    const props = {
      name: 'someDate',
      value: '1981-07-26',
      errors: [ 'does not look right', 'smells funny' ],
    }
    const wrapper = shallow(<DateField {...props} />)
    expect(wrapper.find('.error')).toHaveLength(0)
  })

  it('should display errors if not touched but forceErrorDisplay is true', () => {
    const props = {
      name: 'someDate',
      value: '1981-07-26',
      errors: [ 'does not look right', 'smells funny' ],
      forceErrorDisplay: true,
    }
    const wrapper = shallow(<DateField {...props} />)
    expect(wrapper.find('.error')).toHaveLength(2)
  })

  it('should call onChange when the input changes', () => {
    const props = {
      name: 'someDate',
      value: '1981-07-26',
      onChange: jest.fn(),
    }
    const wrapper = shallow(<DateField {...props} />)
    wrapper.find(DatePicker).prop('onChange')(moment('2014-09-25', DateStorageFormat))
    expect(props.onChange).toHaveBeenCalledWith('2014-09-25')
  })

  it('should call onFocus when the input is focused', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      onFocus: jest.fn(),
    }
    const wrapper = shallow(<DateField {...props} />)
    wrapper.find(DatePicker).prop('onFocus')({ target: { name: 'someInput' } })
    expect(props.onFocus).toHaveBeenCalledWith('someInput')
  })

  it('should focus the input when the focus prop is passed', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      focus: 'someInput',
    }
    const wrapper = shallow(<DateField {...props} />)
    expect(wrapper.find(DatePicker).prop('autoFocus')).toBe(true)
  })
})
