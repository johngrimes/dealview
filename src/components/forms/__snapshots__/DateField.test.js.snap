// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField should be tolerant of an empty value prop 1`] = `
<div
  className="control-group"
>
  <label
    htmlFor="someDate"
  >
    Some Date
  </label>
  <t
    autoFocus={false}
    className=""
    dateFormat="D MMM YYYY"
    dateFormatCalendar="MMMM YYYY"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    monthsShown={1}
    name="someDate"
    onBlur={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    popoverAttachment="top left"
    popoverTargetAttachment="bottom left"
    popoverTargetOffset="10px 0"
    showYearDropdown={true}
    tetherConstraints={
      Array [
        Object {
          "attachment": "together",
          "to": "window",
        },
      ]
    }
    utcOffset={600}
    withPortal={false}
  />
</div>
`;

exports[`DateField should be tolerant of an undefined value prop 1`] = `
<div
  className="control-group"
>
  <label
    htmlFor="someDate"
  >
    Some Date
  </label>
  <t
    autoFocus={false}
    className=""
    dateFormat="D MMM YYYY"
    dateFormatCalendar="MMMM YYYY"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    monthsShown={1}
    name="someDate"
    onBlur={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    popoverAttachment="top left"
    popoverTargetAttachment="bottom left"
    popoverTargetOffset="10px 0"
    showYearDropdown={true}
    tetherConstraints={
      Array [
        Object {
          "attachment": "together",
          "to": "window",
        },
      ]
    }
    utcOffset={600}
    withPortal={false}
  />
</div>
`;

exports[`DateField should display errors if touched 1`] = `
<div
  className="control-group"
>
  <label
    htmlFor="someDate"
  >
    Some Date
  </label>
  <t
    autoFocus={false}
    className=""
    dateFormat="D MMM YYYY"
    dateFormatCalendar="MMMM YYYY"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    monthsShown={1}
    name="someDate"
    onBlur={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    popoverAttachment="top left"
    popoverTargetAttachment="bottom left"
    popoverTargetOffset="10px 0"
    selected={"1981-07-25T14:00:00.000Z"}
    showYearDropdown={true}
    tetherConstraints={
      Array [
        Object {
          "attachment": "together",
          "to": "window",
        },
      ]
    }
    utcOffset={600}
    withPortal={false}
  />
</div>
`;

exports[`DateField should render 1`] = `
<div
  className="control-group"
>
  <label
    htmlFor="someDate"
  >
    Some Date
  </label>
  <t
    autoFocus={false}
    className=""
    dateFormat="D MMM YYYY"
    dateFormatCalendar="MMMM YYYY"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    monthsShown={1}
    name="someDate"
    onBlur={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    popoverAttachment="top left"
    popoverTargetAttachment="bottom left"
    popoverTargetOffset="10px 0"
    selected={"1981-07-25T14:00:00.000Z"}
    showYearDropdown={true}
    tetherConstraints={
      Array [
        Object {
          "attachment": "together",
          "to": "window",
        },
      ]
    }
    utcOffset={600}
    withPortal={false}
  />
</div>
`;
